File: week-2/valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    const hashtable = {};
    for (let i=0; i<board.length; i++) {
        for (let j=0; j<board[i].length; j++) {
            if (board[i][j] !== '.' && !checkBoard(i, j, board, hashtable)) {
                return false
            }
        }
    }
    
    return true;
};

const checkBoard = (i, j, board, hashtable) => {
    const box_row = Math.floor(i/3);
    const box_col = Math.floor(j/3);
    
    const str_x = `${board[i][j]} in row ${i}`;
    const str_y = `${board[i][j]} in col ${j}`;
    const box = `${board[i][j]} in box ${box_row}${box_col}`;
    
    if (hashtable.hasOwnProperty(str_x) || hashtable.hasOwnProperty(str_y) || hashtable.hasOwnProperty(box))
        return false;
    hashtable[str_x] = hashtable[str_y] = hashtable[box] = true;
    
    return true;

}

File: week-2/valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        hashtable = {}
        for i in range(len(board)):
            for j in range(len(board[i])):
                if board[i][j] != '.':
                    if not self.checkValue(i, j, board, hashtable):
                        return False
                    
        return True
    
    def checkValue(self, i, j, board, hashtable):
        box_row = i//3
        box_col = j//3
        
        str_x = str(board[i][j]) + ' in row ' + str(i)
        str_y = str(board[i][j]) + 'in col ' + str(j)
        box = str(board[i][j]) + 'in box ' + str(box_row) + str(box_col)
        
        if str_x in hashtable or str_y in hashtable or box in hashtable:
            return False
        
        hashtable[str_x] = True
        hashtable[str_y] = True
        hashtable[box] = True
        
        return True
